### ISSUE TITLE
r0.2 Errata

### ISSUE THREAD

**Issue by joshajohnson** [joshajohnson/ecp5-mini#2]
This issue will document any errata / areas for improvement for future versions.

Note, whilst the board has passed the smoke test, it has not been fully brought up so this list is subject to growing.

- [ ] USB CC pulldown resistors are in the way of the SD card, card can still be inserted but is less than ideal.
- [ ] PD37+- diff pair is not routed correctly under the SD card.
- [ ] C510 is too close to C514
- [X] C14 and R2 should have their positions swapped in layout.

**Comment by giggiu16** [comment]
Hey @joshajohnson! For a very big coincidence I assembled the r0.2 simultaneously with you! 😄 
Just had the time to perform (a successful!) smoke test and an initial JTAG chain scan (w/ openocd& ftdi/jlink) but no devices were discovered. Do you managed to program the ECP5?
Thank you!

**Comment by joshajohnson** [comment]
Hey @giggiu16, I was able to program both of the r0.2 boards I assembled. I used [ecpprog](https://github.com/gregdavill/ecpprog) with a FT232. 

Let me know how you go / if there is anything I can do to help!

**Comment by giggiu16** [comment]
@joshajohnson -- thank you so much!
Just changed to FT232 + ecpprog combination and...
```
init..
IDCODE: 0x21111043 (LFE5U-12)
ECP5 Status Register: 0x06400000
flash ID: 0xEF 0x40 0x18 0x00
Bye.
```

I confirm also that the bootloader is working correclty even on r0.2:
```
Found DFU: [1d50:614b] ver=0005, devnum=2, cfg=1, intf=0, path="20-4", alt=2, name="Bootloader", serial="e46050640f354126"
Found DFU: [1d50:614b] ver=0005, devnum=2, cfg=1, intf=0, path="20-4", alt=1, name="RISC-V firmware", serial="e46050640f354126"
Found DFU: [1d50:614b] ver=0005, devnum=2, cfg=1, intf=0, path="20-4", alt=0, name="ECP5 bitstream", serial="e46050640f354126"
```

Thank you so much!

**Comment by joshajohnson** [comment]
@giggiu16 Nice work! If you are in the position to share what you are using it for I'd be keen to hear, otherwise have fun!

**Comment by giggiu16** [comment]
@joshajohnson for sure: no specific use case in mind for now. I'll use your board to continue learning FPGA programming, I didn't wanted to be limitated in GPIO pins too much comparing other "low-cost" solutions. I like also assemblying electronic prototypes and I used your board as first BGA assembly test 😄. Next 2 boards in queue, and one is with the 25F variant 😉.

**Comment by joshajohnson** [comment]
@giggiu16 Awesome!

One thing I should say is that the 12F and the 25F are the same silicon, but with a different IDCODEs. If you use the open tools you'll be able to use all 25K LUTs on the 12F!

**Comment by amitvasudevan** [comment]
Hi @joshajohnson and @giggiu16! Nice work on the ECP5-mini board. How did you solder the BGA on the PCB? Did you just use a hot air gun/flux combo or was it something more sophisticated? Thanks!

**Comment by joshajohnson** [comment]
Hi @amitvasudevan, I'd suggest reflowing it along with the rest of the components. You can do it with hot air but getting even heating is a bit more challenging.

**Comment by joshajohnson** [comment]
Closed on accident

**Comment by amitvasudevan** [comment]
Thanks for the response @joshajohnson !

I am a newbie when it comes down to soldering so please forgive my naive questions below :)

- From what I read up on reflow seems like a three step process: (a) obtain stencil for PCB, (b) apply solder paste over stencil, and (c) place the pcb in the reflow oven. Correct?
- I am planning to order the PCB via OSHPark; do they provide stencils?
- Also I noticed that your PCB design has components soldered on both the top and bottom. How does that work in the reflow oven? Do you do one side first and then the other? If so, won't the already soldered components on the side down reflow and dislodge?
- I am planning on assembling this at home and was looking for reflow ovens on amazon and it seems like there are units available for ~$250 or so. Is there a specific unit you would recommend? Also I am guessing the temperature will depend on the solder paste being used correct?

Thanks for your time and help! Can't wait to get started...

**Comment by joshajohnson** [comment]
- Correct

- OSHStencils (no relation to OSHPark) provide stencils.

- Yes, you reflow the bottom side then the top side. There is no issues with the bottom side falling off on the second time through the oven as they are all lightweight.

- Esden has a good summary of modifying an oven [here](https://forum.1bitsquared.com/t/upgrading-a-t962a-reflow-oven/126). I have a T962 (not T962A) and it's okay, but I'd buy the higher end option next time.

- Yes you will want to play around with a few boards / temperature profiles before throwing your first actual board in there to see what works best.

**Comment by amitvasudevan** [comment]
Thanks @joshajohnson! This is very helpful. I looked at the T962/A modifications and found the thermocouple interface board is no longer available from 1bitsquared. Is that an important component or can I still work without it?

**Comment by joshajohnson** [comment]
You can get away without it, I don't have one in my machine.
Just make sure you add the [cold junction compensation DS18B20](https://github.com/UnifiedEngineering/T-962-improvements/wiki/Dallas-18B20-for-cold-junction-mod) in the controller board to get a bit closer with the temperature.

**Comment by amitvasudevan** [comment]
Got it! Would you be willing to share details on the solder paste you used and the corresponding T962 profile that worked for you? I am planning to use lead-free solder paste (just to be less toxic) and it seems like it could work with the SAC305 lead-free paste as described [here](https://www.youtube.com/watch?v=Hkl6rzhMPHY). Thanks again for your time and help!

**Comment by joshajohnson** [comment]
I use leaded T4 paste so the profile won't be too useful.
Main thing to keep an eye out for is that any fast ramp rates will cause the oven to lag as it can't keep up with the fast changes in temperature, so pick a slow profile and run it a few times to see how the oven responds.

I've had my oven for 4 years now and as long as I put a part less than 80mm x 80mm in the center of the board it will solder without too many issues, but it took a while to get to that point.

**Comment by amitvasudevan** [comment]
Thanks @joshajohnson! I have placed my order for the T962 and it should get here next week. Also, I found a lead-free solder paste with Tin/Bismuth alloy from Chipquik. If you look at the datasheet [here](http://www.chipquik.com/datasheets/SMDLTLFP.pdf), it seems to have a pretty smooth reflow curve (about 1C temp rise per second). Is that a slow enough profile for the oven you think?

**Comment by joshajohnson** [comment]
Ah sorry I might not have been clear - you can use any solder paste but just need to make sure the ramp rate on your oven is slow.
Mine has two options - one shaped like a triangle and another that has a long soak profile before peaking, and I use the latter as it can't match the sharp profile.

**Comment by Oscilllator** [comment]
Hi,

Did anyone manage to make use of the RAM on this board? I'm doing my first ECP5 board and am looking for designs to work from. RAM is a nice-to-have for my design so it would be good to have a known good implementation to work from.

**Comment by joshajohnson** [comment]
Hi Dave,

Hyperram has been confirmed to work via a memtest in Litex.

Code is here:

https://github.com/joshajohnson/ecp5-mini-projects/tree/master/litex

Thanks

-------- Original Message --------

**Comment by CCampbell71** [comment]
@joshajohnson I'm looking at assembling one! But the 25 vs 12... you mentioned the IDCODEZ is different, will I have issues flashing and using the dfu bootloader?

**Comment by joshajohnson** [comment]
Hi @CCampbell71, you can fit whatever part you want. Just make sure this part of the makefile is correct for the IC.
https://github.com/joshajohnson/ecp5-mini-projects/blob/master/verilog/common.mk#L3

**Comment by CCampbell71** [comment]
Thanks @joshajohnson! Ok so I got the pcb assembled, I used your design for the FPGA portion including usb, buttons, jtag, spi etc. exactly the same. I am now unable to connect via jtag. I tried ECPPROG and open fpga loader, and am using a TIGARD (Ftdi) cable. Verified all power banks are correct and DONE pin pulls low indicating it moved into initialization. But I cannot get it to recognize the fpga via jtag. Any recommendations for the first flash? It has not been configured before

**Comment by joshajohnson** [comment]
It's been so long since I flashed an ECP5 I'm struggling to remember what it looks like @CCampbell71!

Assuming you are not getting any readback of an IDCODE at all?
All I can think of is a dodgy connection between Tigard (correct pins to each spot? tight fit in holes?) or bad soldering on the BGA.
Otherwise maybe it's a driver / udev issue? Can you try a different device and see if you can find that on JTAG?

Sorry for the non answer, let me know how you go.

**Comment by CCampbell71** [comment]
Yes, I've triple checked all connections and used openfogaloader/ ecpprog and native diamond software. Still cannot get id code :/. Question: when you have a fresh brand new board that's never been configured, you simply power it up and connect JTAG? I do see behavior of the DONE pin led come on and then goes off indicating the FPGA moved on to the initialization phase. But doesn't configure (obviously) also tried straight up flashing the spi chip (Winbond) and I'm getting hit with the write protection stuff. Working through that with various software/programmers as well 😄

**Comment by CCampbell71** [comment]
I did notice the hardware checklist callls for 4.kohm pullups on most of the jtag pins. I didn't do this but looks like you didn't either so I figured either was a non issue. TCK is pulled down via 10kohm

**Comment by joshajohnson** [comment]
Yeah I think you just connect JTAG and send it, IIRC you should always be able to get the IDCODE irrespective of the device state.
Pin strapping sounds fine to me. Sorry for not being much use.
